python/lsst/ts/salobj/async_s3_bucket.py

```diff
@@ -0,0 +1,130 @@
+# This file is part of ts_salobj.
+#
+# Developed for the LSST Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["AsyncS3Bucket"]
+
+import asyncio
+import io
+
+import boto3
+import botocore
+
+
+class AsyncS3Bucket:
+    """Asynchronous interface to Amazon Web Services s3 buckets.
+
+    Parameters
+    ----------
+    name : `str`
+        Name of bucket.
+    """
+    def __init__(self, name):
+        s3 = boto3.resource("s3")
+        self.name = name
+        self.bucket = s3.Bucket(name)
+
+    async def upload(self, fileobj, key, callback=None):
+        """Upload a file-like object to the bucket.
+
+        Parameters
+        ----------
+        fileobj : file-like object
+            File-like object that can be read as *binary* data.
+        key : `str`
+            Name to use for the file in the bucket.
+        callback : callable (optional)
+            Function to call with updates while writing. The function receives
```

Comment:
It might be worth mentioning that this is passed on through to the s3 `upload_fileobj` method, if someone is curious how it behaves in detail (since you have no control over that).

Reply:
OK. I have added it to the Notes for upload and download. I will say that the boto3 documentation is poor, and in this case it says almost nothing about how these callbacks are handled. I had to work out the details that I mention in the doc string by myself, and hope that they will not change in the future.